src/realm/object-store/impl/set_notifier.cpp

```diff
@@ -63,6 +63,22 @@ bool SetNotifier::do_add_required_change_info(TransactionChangeInfo& info)
     info.lists.push_back({m_table, m_obj.value, m_col.value, &m_change});
 
     m_info = &info;
+
+    // When adding or removing a callback, the related tables can change due to the way we calculate related tables
+    // when key path filters are set, hence we need to recalculate every time the callbacks are changed.
+    // We only need to do this for lists that link to other lists. Lists of primitives cannot have related tables.
+    util::CheckedLockGuard lock(m_callback_mutex);
+    if (m_did_modify_callbacks && m_type == PropertyType::Object) {
+        m_related_tables = {};
+        auto& set = static_cast<LnkSet&>(*m_set);
+        recalculate_key_path_arrays();
+        DeepChangeChecker::find_filtered_related_tables(m_related_tables, *(set.get_target_table()),
+                                                        m_key_path_arrays);
+        // We deactivate the `m_did_modify_callbacks` toggle to make sure the recalculation is only done when
+        // necessary.
+        m_did_modify_callbacks = false;
+    }
```

Comment:
I notice this logic appears in a few places. Is it possible to extract it to a shared function?

Reply:
Fixed with c00fd071b.